### DISCUSSION TITLE
Serialize form input values as array

### DISCUSSION THREAD

**Discussion by yulafezmesi** [remix-run/remix#1615] [Proposals]
Basically, we can access single input values with name attributes on the server-side. That would be awesome if we can access dynamic nested array values over the server-side as well. 

So for example:

```html
<input value="Hey" type="text" name="items[]" />
<input value="Remix" type="text" name="items[]" />
<input value="Awesome" type="text" name="items[]" />
```

Expected output on the server-side: 

```javascript
formData: {
        //...other values
        items: [
            "Hey",
            "Remix",
            "Awesome"
        ]
    }
```

**Comment by ccssmnn** [comment]
Is this a native web forms functionality? If yes, it should just work. Remix goes all-in into web standards, complementing them with features to provide a modern user experience. I believe, adding non-standard behavior to input elements will not a feature, remix will get. 

However, you can still submit a request with client-side js and put whatever values you want in there.

**Reply by yulafezmesi** [reply]
Actually, this is not directly related to web forms standards. It depends on how the backend serializes/parses your form values. [Express.js](https://expressjs.com/en/resources/middleware/body-parser.html#extended) already handles this parsing if you pass the `{ extended: true }` parameter to the parser. 

I've created a small POC for that. Submit native HTML form and see how to parse it.

https://stackblitz.com/edit/expressjs-1fwyjw

**Reply by ccssmnn** [reply]
The link does not work. Can you share the code?

**Reply by yulafezmesi** [reply]
Ups, sorry.

https://stackblitz.com/edit/expressjs-1fwyjw

**Reply by ccssmnn** [reply]
This looks like a syntax for input field names, to be parsed as more complex objects. Express uses qs under the hood.

You should be able to read the query string and use [qs](https://www.npmjs.com/package/qs#readme) manually.

**Comment by rvanlaarhoven** [comment]
There are some examples on how to deal with this in https://github.com/remix-run/remix/discussions/1541

**Comment by ryanflorence** [comment]
https://remix.run/docs/en/v1/pages/faq#how-can-i-have-structured-data-in-a-form

```jsx
<input value="Hey" type="text" name="items" />
<input value="Remix" type="text" name="items" />
<input value="Awesome" type="text" name="items" />

// ...
formData.getAll("items")
// ["Hey", "Remix", "Awesome"]
